prov/lnx/src/lnx_init.c

```diff
@@ -251,300 +264,310 @@ static int lnx_cache_info(struct dlist_entry *head,
+static struct fi_info *
+lnx_filter_info2(struct fi_info *fi, char *domain)
 {
-	struct fi_info *fi = NULL, *head = NULL, *prev = NULL;
-	struct lnx_fi_cache_entry *e;
-	int rc, size;
-
-	/* we need at least 2 providers to link */
-	rc = lnx_get_cache_meta(&lnx_fi_info_cache, &size);
-	if (rc || size < 2)
-		return -FI_ENODATA;
-
-	rc = gen_links(&lnx_fi_info_cache, &lnx_links, size);
-	if (rc)
-		return rc;
-
-	/*
-	 * 1. Iterate over the links and create a linked list of fi_infos
-	 *    each fi_info in the list represents one of the links
-	 * 2. Have metadata associated with each fi_info to refer back to
-	 *    an entry in the lnx_links cache.
-	 * 3. When the application selects one of these fi_infos, we can
-	 *    then find the appropriate link in the cache and be able to
-	 *    create the underlying core providers correctly.
-	*/
-	dlist_foreach_container(&lnx_links, struct lnx_fi_cache_entry, e,
-				entry) {
-		rc = lnx_form_info(e->fi, &fi);
-		if (rc)
-			goto err;
-
-		if (prev) {
-			prev->next = fi;
-			prev = fi;
-		} else {
-			prev = fi;
-			head = fi;
+	struct fi_info *itr, *next, *discard = NULL, *discard_tail, *prev;
+
+	prev = itr = fi;
+	while (itr) {
+		if (strcmp(itr->domain_attr->name, domain)) {
+			if (!discard)
+				discard = itr;
+			if (itr == fi) {
+				fi = itr->next;
+				prev = fi;
+				next = prev;
+			} else {
+				prev->next = itr->next;
+				next = prev->next;
+			}
+			itr->next = NULL;
+			discard_tail->next = itr;
+			discard_tail = itr;
+			itr = next;
+			continue;
 		}
+		prev = itr;
+		itr = itr->next;
 	}
+	if (discard)
+		fi_freeinfo(discard);
 
-	*info = head;
-
-	return FI_SUCCESS;
-
-err:
-	if (fi)
-		fi_freeinfo(fi);
-	lnx_free_info_cache(&lnx_fi_info_cache, false);
-	lnx_free_info_cache(&lnx_links, true);
-
-	return -FI_ENODATA;
+	return fi;
 }
+*/
 
-int lnx_getinfo_helper(uint32_t version, char *prov, struct fi_info *lnx_hints)
+int lnx_getinfo_helper(uint32_t version, char *prov, char *domain,
+		       struct fi_info *lnx_hints, bool new_prov,
+		       struct dlist_entry *link)
 {
 	int rc;
-	char *orig_prov_name = NULL;
+	char *orig_prov_name = NULL, *orig_dom_name = NULL;
 	struct fi_info *core_info;
-	uint64_t caps, mr_mode;
+	uint64_t caps;
 	bool shm = false;
 
 	caps = lnx_hints->caps;
-	mr_mode = lnx_hints->domain_attr->mr_mode;
 
-	if (lnx_hints->fabric_attr->prov_name) {
-		orig_prov_name = lnx_hints->fabric_attr->prov_name;
-		lnx_hints->fabric_attr->prov_name = NULL;
-	}
+	orig_prov_name = lnx_hints->fabric_attr->prov_name;
+	orig_dom_name = lnx_hints->domain_attr->name;
 
 	lnx_hints->fabric_attr->prov_name = prov;
+	lnx_hints->domain_attr->name = domain;
+
 	if (!strncmp(prov, "shm", 3)) {
 		shm = true;
-		/* make sure we get the correct shm provider */
-		lnx_hints->caps &= ~(FI_REMOTE_COMM | FI_LOCAL_COMM);
-		lnx_hints->caps |= FI_HMEM;
-		lnx_hints->domain_attr->mr_mode |= (FI_MR_VIRT_ADDR | FI_MR_HMEM
-						| FI_MR_PROV_KEY);
+		lnx_hints->caps &= ~(FI_REMOTE_COMM);
```

Comment:
don't know if anyone does this but technically it can show up in hints->domain_attr->caps as well

Reply:
done